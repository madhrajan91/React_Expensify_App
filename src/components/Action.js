import React from 'react'
//Action functional vs class 
const Action = (props) => {
    return (
        <div>
                <button onClick={props.handlePick}
                    disabled={!props.hasOptions}
                >
                    What should I do
                </button>
            </div>
    )
}
/*
class Action extends React.Component {
    constructor(props) {
        super(props);
        this.hasOptions = this.props.hasOptions;
    }
    render() {
        return (
            <div>
                <button onClick={this.props.handlePick}
                disabled={!this.hasOptions}
                >What should I do</button>
            </div>
        );
    }
}
*/
export default Action;