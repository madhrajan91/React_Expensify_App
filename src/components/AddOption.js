import React from 'react';

// add Option
export default class AddOption extends React.Component {
    constructor(props) {
        super(props)
        this.handleSubmitOption = this.handleSubmitOption.bind(this);
        this.state = {
            "error": undefined
        }
    }
    handleSubmitOption (e) {
        e.preventDefault();
        const option = e.target.elements.option.value.trim();

        // passing option upstream since handleAddOption is coming from
        // the <InDecisionApp />
        const error = this.props.handleAddOption(option);
        if (error) {
            // this.setState ( () => {
            //     return {
            //         "error": error
            //     };
            // });

            // or
            this.setState (() => ({"error": error}))
        }
        // if (!error) {
        //     e.target.elements.option.value = "";
        // }
        //}
    }
    render() {
        return (
            <div>
                {this.state.error && <p>ERROR {this.state.error}</p>}
                <form onSubmit={this.handleSubmitOption}>
                    <input type="text" name="option" />
                    <button> Add Option </button>
                </form>
            </div>
        );
    }
}
