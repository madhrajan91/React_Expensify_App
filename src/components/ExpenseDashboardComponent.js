import React from 'react';
import ConnectedExpenseListComponent from './ExpenseListComponent';


const ExpenseDashboardComponent = () => (
    <div>
        This is from my dashboard component
        <ConnectedExpenseListComponent />
    </div>
)

export default ExpenseDashboardComponent;