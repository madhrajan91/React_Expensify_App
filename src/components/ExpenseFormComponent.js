import React from 'react';
import moment from 'moment';

import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css'

const now = moment();
console.log(now.format('MMM Do, YYYY'))

export default class ExpenseFormComponent extends React.Component {
    // use local state instead of redux store since this is just local to this component


    constructor(props) {
        super(props);
        this.state = {
            description:props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e)=> {
        const description = e.target.value;
        this.setState(()=> {
            return {
                description //same as description:description
            }
        })
    }
    onAmountChange = (e)=> {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(()=> {
                return {
                    amount
                }
            })
        }
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        //e.persist() if you want to use e.target.value directly below
        this.setState( () => {
            return {
                note
            }
        })
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState({createdAt})
        }
    }

    onFocusChange = ({focused}) => {
        this.setState(()=> ({calendarFocused:focused}))
    }

    onSubmit = (e) => {
        if (!this.state.description || !this.state.amount) {
            //set error
            this.setState(() => ({
                error: 'Please enter valid descripton or amount'
            }))
        }
        else {
            this.setState(() => ({
                error: ''
            }))
            this.props.onSubmit({
                description:this.state.description,
                amount: parseFloat(this.state.amount,10) * 100,
                createdAt: (this.state.createdAt).valueOf(),
                note: this.state.note

            })
        }
        e.preventDefault();
    }
    render() {
       return(
            <div>
             ExpenseForm
             {this.state.error !== '' && <p>{this.state.error}</p>}
             <form onSubmit={this.onSubmit}>
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}         
                />
                <SingleDatePicker
                    date = {this.state.createdAt}
                    onDateChange= {this.onDateChange}
                    focused = {this.state.calendarFocused}
                    onFocusChange = {this.onFocusChange}
                    numberOfMonths = {1}
                    isOutsideRange = {()=>false}
                />
                <textarea
                    placeholder="note" value={this.state.note} onChange={this.onNoteChange}>
                </textarea>
                <button>Add Expense</button>
             </form>
            </div>
       );
    }
}