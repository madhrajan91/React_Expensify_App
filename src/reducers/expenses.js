// Expenses REducer
const expensesReducerDefaultState = []
const expensesReducer = (state= expensesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense); // don't use push since push adds to the original while concat returns a new object (Could   USE SPREAD instead)
            return [... state, action.expense]; // SPREAD
        case 'REMOVE_EXPENSE':
            const expenses = state.filter((expenseToRemove) => {
                return (expenseToRemove.id != action.id);
            })
            return expenses;
        case 'EDIT_EXPENSE':
            return state.map((expense)=> {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else {
                    return expense;
                }
            })
        default:
            return state;
    }
}

export default expensesReducer;