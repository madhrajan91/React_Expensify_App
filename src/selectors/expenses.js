const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    const filteredExpenses = expenses.filter((expense) => {
       console.log(text)
       const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
       const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
       const textMatch = text === '' || expense.description.toLowerCase().includes(text.toLowerCase()) ;

       return startDateMatch && endDateMatch && textMatch;
   })
       console.log(sortBy)
       return filteredExpenses.sort((a,b) => {
           if (sortBy === 'date') {
               return a.createdAt > b.createdAt?1:-1;
           }
           else {
               return a.amount > b.amount?1:-1;
           }
       })

   //return expenses;
}

export default getVisibleExpenses;