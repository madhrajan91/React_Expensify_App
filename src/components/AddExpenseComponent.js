import React from 'react';
import {connect} from 'react-redux'

import ExpenseFormComponent from './ExpenseFormComponent'
import {addExpense} from '../actions/expenses'

const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    };
}

export class AddExpenseComponent extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/')
    }
    render() {
        return (
             <div>
                Add expense
                <ExpenseFormComponent
                    onSubmit = {this.onSubmit}
                />
            </div>
        )
    }
}
/*
const AddExpenseComponent = (props) => (
    <div>
        Add expense
        <ExpenseFormComponent
         onSubmit = {(expense) => {

            console.log(expense)
            props.onSubmit(expense);
            //props.dispatch(addExpense(expense));
            props.history.push('/')
         }} />
    </div>
)
*/

export default connect(undefined, mapDispatchToProps)(AddExpenseComponent);