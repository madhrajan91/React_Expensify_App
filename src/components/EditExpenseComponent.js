import React from 'react';

import {connect} from 'react-redux';

import ExpenseFormComponent from './ExpenseFormComponent';
import {editExpense} from '../actions/expenses'
import {removeExpense} from './../actions/expenses';

const EditExpenseComponent = (props) => {
    console.log(props);
    // query strings 
    // hash
    console.log(props.location.search);
    console.log(props.location.hash);
    return (
        <div>
            Edit  with id of {props.match.params.id}
            <ExpenseFormComponent onSubmit={(expense)=> {
                console.log(expense.id);
    
                props.dispatch(editExpense(props.match.params.id, expense));
                props.history.push('/')
            }}
            expense={props.expense}
            />
            <button onClick= {(e) => {
                props.dispatch(removeExpense(props.expense));
                console.log('removed');
                props.history.push('/')
            }}>Remove</button>
        </div>
    )
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            if (expense.id === props.match.params.id){
                return true;
            }
            else {
                return false;
            }
        })
    }
}

export default connect(mapStateToProps)(EditExpenseComponent);