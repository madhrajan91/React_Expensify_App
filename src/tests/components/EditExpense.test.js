import React from 'react';
import {shallow} from 'enzyme';
import expenses from './../fixtures/expenses';
import { EditExpenseComponent } from '../../components/EditExpenseComponent';


let editExpense, removeExpense, history, wrapper, match;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn()
    history = {push: jest.fn() }
    match = {
        params: {
            id: expenses[1].id
        }
    }
    wrapper = shallow(<EditExpenseComponent expense={expenses[1]} editExpense={editExpense} removeExpense={removeExpense} history={history} match={match}/>);

})

test('should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
})

test('should handle edit expense correctly', () => {
    wrapper.find('ExpenseFormComponent').prop('onSubmit')(expenses[1]);
    expect(wrapper).toMatchSnapshot();
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
})

test('should handle delete expense correctly', () => {

    wrapper.find('button').at(0).simulate('click', {});
    expect(wrapper).toMatchSnapshot();
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[1]);
})