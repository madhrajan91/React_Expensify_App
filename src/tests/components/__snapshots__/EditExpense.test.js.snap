// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle delete expense correctly 1`] = `
<div>
  Edit  with id of 
  2
  <ExpenseFormComponent
    expense={
      Object {
        "amount": 20,
        "createdAt": 5000,
        "description": "Gum bill1",
        "id": 2,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should handle edit expense correctly 1`] = `
<div>
  Edit  with id of 
  2
  <ExpenseFormComponent
    expense={
      Object {
        "amount": 20,
        "createdAt": 5000,
        "description": "Gum bill1",
        "id": 2,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should render edit expense page correctly 1`] = `
<div>
  Edit  with id of 
  2
  <ExpenseFormComponent
    expense={
      Object {
        "amount": 20,
        "createdAt": 5000,
        "description": "Gum bill1",
        "id": 2,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
